Refactor and update mc_docx_update to accomodate new database changes

maxachis opened the pull request:
## What I did

Add comments and type hinting
Add print statements indicating status of conversion process. 
Light refactoring
Modify which tables are modified and how during update/insertion
Added SQL script for inserting test data (currently only one case)

## Why I did it

Modifying script in accordance with database changes as well as prepping for data loading. 

## Issue #s Addressed

https://trello.com/c/HmsXVeVk/30-script-that-takes-link-to-bill-and-extracts-text-from-bill

## How to test it

1. Will need to setup local Supabase DB on Docker.
2. In addition to setting up schema data, will need to run scripts in `tests/sql/setup_test_docx_etl.sql` 
3.  I used ubuntu in VM via PyCharm. We probably need a docker-based way of testing it, but not sure how that would work at this moment.
4. In vm, type `sudo apt install libreoffice`
5. Run `pip install [package]` for the following packages: python-docx, supabase, python-dotenv
6. Run `python src/docx-etl.py`

Will need dev version of DB as well as a handful of test entries created in Revisions table with links to the actual bills. Haven't yet developed that script.

maxachis committed:
Add comments and type hinting
Light refactoring
Modify which tables are modified and how during update/insertion

maxachis commented:
In progress test instructions:
1.  I used ubuntu in VM via PyCharm. We probably need a docker-based way of testing it, but not sure how that would work at this moment.
2. In vm, type `sudo apt install libreoffice`
3. Run `pip install [package]` for the following packages: python-docx, supabase, python-dotenv
4. Run `python src/docx-etl.py`

maxachis committed:
Add print statements, bug-fix, and docx_etl test data.
Added clarifying print statements to show status -- is this desirable?
Corrected bugs
Setup sql for inserting test data.

mikeandike523 commented on src/docx-etl.py:
```diff
@@ -4,58 +4,94 @@
     response = requests.get(bill_link, stream=True)
     output_file_name = "bill.doc"
-    with open(output_file_name, "wb") as f: #add try catch here and return empty filname if it fails
+    with open(output_file_name, "wb") as f:  # add try catch here and return empty filename if it fails
```
1. Shouldn't return the empty filename, should return `str` or `None`, i.e. `Optional[None]`
2. I recommend putting it in a temp folder instead of the main repo folder, like "temp/bill.doc", this way can can add "temp" to `.gitignore`

maxachis replied:
Nice catch on both; addressing both in my upcoming commit.

mikeandike523 commented on src/docx-etl.py:
```diff
@@ -4,58 +4,94 @@
+    # TODO: find a way to determine when the actual bill text starts
+    doc = Document(bill_path)
+    output_text = ""
+    for paragraph in doc.paragraphs:
```
Please test to see if each paragraph ends with a newline (`\n` or `\r\n`), if not, we need to insert it after reach paragraph

maxachis replied:
Good catch; added. This might not be terribly important since we'll be mostly using the raw text for analysis, but adding the line breaks doesn't hurt and can make it a bit more readable.

mikeandike523 commented on tests/sql/setup_test_docx_etl.sql:
```diff
@@ -0,0 +1,18 @@
+insert into "Bills" (legislative_id, legislative_session, bill_number, session_type, chamber)
```
consider making this idempotent, using something like `insert ignore` command. As it stands, if in any situation the .sql query is run more than once, an error will be thrown

maxachis replied:
Postgres SQL doesn't have an insert ignore command, but it does have something similar ("On conflict (value) do nothing"), so I've used that instead.

mikeandike523 commented on tests/sql/setup_test_docx_etl.sql:
```diff
@@ -0,0 +1,18 @@
+   'HOUSE'
+)
+
+insert into "Revisions" (revision_guid, printer_no, full_text_link, publication_date, description, bill_internal_id)
```
consider making this idempotent, using something like `insert ignore` command. As it stands, if in any situation the .sql query is run more than once, an error will be thrown

maxachis replied:
See my above comment re: insert ignore

maxachis committed:
Modify setup_text_docx_etl to not throw an error if an insert of the same revision_guid is attempted twice.

maxachis committed:
Minor revisions to docx-etl and .gitignore:
- bill.doc now downloaded to and converted within an explicit "temp" directory
- Newline character added at the end of each paragraph in extract_law_text_from_docx

Add .temp/ directory to gitignore

maxachis commented:
I'm going to go ahead and merge this without waiting for approvals. My rationale is that I implemented most of the changes people suggested and I didn’t want to busy up the next meeting by reviewing the same code. If people find there are issues with the changes merged, we can always make changes to it later on and, in a pinch, revert the commit. 

I’m also revising/formalizing pull request standards so that we can get PRs moving faster and avoiding stalls.
